### DISCUSSION TITLE
Immutability by default

### DISCUSSION THREAD

**Discussion by marcus-sa** [DanielXMoore/Civet#1552] [Ideas]
One of the things I hate the most about TypeScript is that everything is mutable by default.
I wish there was an option that made everything immutable by default at compile time.
It'd be great if it something similar could be implemented in Civet.

**Comment by edemaine** [comment]
Do you have a specific proposal? Mutable is such a deep aspect of the runtime that I doubt this is really doable in something like Civet. But if you have specific ideas (freezing everything all of the time?), happy to discuss.

**Reply by marcus-sa** [reply]
I'm not talking about runtime. I'm talking about compile time. From what I've understood Civet compiles to TypeScript. Adding `readonly` to all types would be one way to solve it.

**Reply by marcus-sa** [reply]
However freezing everything all of the time sounds like a good idea for runtime immutability.

**Comment by STRd6** [comment]
One feature we could add that may get us 70% of the way there is to add a compiler flag to default `:=` declarations to `as const`. There's still a lot more to do for full immutability but this could at least cover the most common cases and we could expand it bit by bit as we find other areas.

**Reply by marcus-sa** [reply]
`as const` only work for literal types.
